File: day2/py/main.py
def checkgame(line):
    line = line.rstrip('\n')
    splits = line.split(':')
    id = int(splits[0].split(' ')[1])
    sets = splits[1].split(';')
    for set in sets:
        blue = green = red = 0
        for part in set.split(','):
            elems = part.split(' ')
            match elems[2]:
                case "blue":
                    blue += int(elems[1])
                case "green":
                    green += int(elems[1])
                case "red":
                    red += int(elems[1])    
        if red > 12 or green > 13 or blue > 14:
            return 0
    return id 

def powergame(line):
    blue = green = red = 1
    line = line.rstrip('\n')
    splits = line.split(':')
    sets = splits[1].split(';')
    for set in sets:
        for part in set.split(','):
            elems = part.split(' ')
            val = int(elems[1])
            match elems[2]:
                case "blue":
                    blue = val if val > blue else blue
                case "green":
                    green = val if val > green else green
                case "red":
                    red = val if val > red else red
    return blue * green * red

lines = open("../input.txt", "r").readlines()

#Part One
sum = 0
for line in lines:
    sum += checkgame(line)
print("part one =", sum)

#Part Two
sum = 0
for line in lines:
    sum += powergame(line)
print("part two =", sum)

File: day2/cpp/main.cpp
#include <cstdio>
#include <fstream>
#include <string>
#include <sstream>
#include <vector>

int playGame(const std::string &line)
{
    int pos = line.find(':') + 1;
    int id = std::stoi(line.substr(line.find(' '), pos), nullptr);
    std::istringstream sets(line.substr(pos, line.size()));
    for (std::string set; std::getline(sets, set, ';'); ) {
        int blue = 0, green = 0, red = 0;
        std::istringstream elems(set);
        for (std::string elem; std::getline(elems, elem, ',') ;) {
            int delim = elem.find_last_of(' ');
            int val = std::stoi(elem.substr(0, delim), nullptr);
            std::string color = elem.substr(delim + 1, elem.size());
            if (color == "blue") {
                blue += val;
            } else if (color == "green") {
                green += val;
            } else {
                red += val;
            }
        }
        if ((red > 12) || (green > 13) || (blue > 14)) {
            return 0;
        }
    }
    return id;
}

int powerGame(const std::string &line)
{
    int blue = 1, green = 1, red = 1;
    int pos = line.find(':') + 1;
    std::istringstream sets(line.substr(pos, line.size()));
    for (std::string set; std::getline(sets, set, ';'); ) {
        std::istringstream elems(set);
        for (std::string elem; std::getline(elems, elem, ',') ;) {
            int delim = elem.find_last_of(' ');
            int val = std::stoi(elem.substr(0, delim), nullptr);
            std::string color = elem.substr(delim + 1, elem.size());
            if (color == "blue") {
                blue = (val > blue) ? val : blue;
            } else if (color == "green") {
                green = (val > green) ? val : green;
            } else {
                red = (val > red) ? val : red;
            }
        }
    }
    return blue * green * red;
}

int main()
{
    std::ifstream input("..\\input.txt");
    std::vector<std::string> lines;
    std::string line;
    while (std::getline(input, line)) {
        lines.push_back(line);
    }

//Part One
    int sum = 0;
    for(const std::string &line : lines) {
        sum += playGame(line);
    }
    printf("part one = %d\n", sum);

//Part Two
    sum = 0;
    for(const std::string &line : lines) {
        sum += powerGame(line);
    }
    printf("part two = %d\n", sum);

    return 0;
}